### PR TITLE
Remove txindex requirement for full nodes

### LINKED ISSUE #527: Consider avoiding the need for `txindex=1`
I'm running into serious performance issues when using `txindex=1` on Bitcoin Core, even without pruning. `btcd` is also taking weeks to do an IBD although I can't tell if that's because of `txindex=1` or some other reason.

It would be sad if users can't (super easily) use their own full node because of this. To [quote](https://botbot.me/freenode/bitcoin-core-dev/2017-12-27/?msg=95099026&page=3) @sipa: "what?" :-)

Can you explain what this index is used for? Maybe I can help brainstorm some ways to avoid that.

### PR DESCRIPTION
**Pull request by wpaulino**
Fixes #527.

So far, `lnd` has required its backend `btcd` and `bitcoind` full nodes to enable the transaction index. It is used to retrieve details for transaction confirmations and/or already spent transactions. This PR aims to remove that requirement by introducing some fallback methods, in the event that the transaction index is not enabled.

When attempting to retrieve these details, we now have the following workflow:

1. We attempt to look up transactions using the transaction index.
2. If the transaction index is not enabled, we attempt to look up transactions by manually scanning the chain's blocks starting from the earliest height where the transaction could have been included in, to the current height.
~~3. In the case of pruned nodes, some of these blocks that we tried to scan above might have been missing because the node discarded them. Therefore, we decide to fetch these blocks from the node's peers and scan them.~~

~~I've also added some unit tests to make sure each part of the workflow above works as expected.~~

~~It's important to get this right, so it would be really helpful if people could spin up this patch locally on one of their test nodes and see how things go.~~

~~This depends on https://github.com/Roasbeef/btcwallet/pull/14, so you'll either need to pull that locally as well or make the changes yourself within the `vendor` directory (very minor).~~

I realize there's a lot of code duplication, it might be worth looking into how to refactor some of these things. That's probably out of scope for this PR though.

### CODE REVIEW

**Comment by wilmerpaulino** [comment]
Looks like to fully support pruned nodes, we'll also have to cover some parts in the `routing/chainview` and `lnwallet/btcwallet` packages.

@Roasbeef, should I open a new PR to address this?

File: chainntnfs/bitcoindnotify/bitcoind.go [hunk 1]
```diff
@@ -486,127 +978,80 @@ type spendCancel struct {
...
-				return nil, err
-			}
-			b.heightMtx.Lock()
-			currentHeight := b.bestHeight
-			b.bestHeight = blockHeight
-			for i := currentHeight; i > blockHeight; i-- {
-				err = b.txConfNotifier.DisconnectTip(uint32(i))
-				if err != nil {
-					return nil, err
+		return &chainntnfs.SpendEvent{
```
Comment (halseth) [inline chainntnfs/bitcoindnotify/bitcoind.go]: This can be moved after the switch, similar to how it was before.

File: chainntnfs/bitcoindnotify/bitcoind.go [hunk 2]
```diff
@@ -716,3 +1161,102 @@ func (b *BitcoindNotifier) RegisterBlockEpochNtfn() (*chainntnfs.BlockEpochEvent
 		}, nil
 	}
 }
+
+// nodePeer is a peer connected to the RPC client's backend coupled with a
+// channel that will retrieve messages from it.
+type nodePeer struct {
```
Comment (halseth) [inline chainntnfs/bitcoindnotify/bitcoind.go]: move to top?

File: chainntnfs/bitcoindnotify/bitcoind.go [hunk 3]
```diff
@@ -716,3 +1161,102 @@ func (b *BitcoindNotifier) RegisterBlockEpochNtfn() (*chainntnfs.BlockEpochEvent
...
+	if err != nil {
+		return nil, err
+	}
+
+	// Allow the peer to start exchanging messages on the established
+	// connection.
+	p.AssociateConnection(conn)
+
+	// Since the peer will reject all messages without establishing a
+	// version handshake first, we make sure it has succeeded.
```
Comment (halseth) [inline chainntnfs/bitcoindnotify/bitcoind.go]: 👍

File: chainntnfs/bitcoindnotify/bitcoind.go [hunk 4]
```diff
@@ -716,3 +1161,102 @@ func (b *BitcoindNotifier) RegisterBlockEpochNtfn() (*chainntnfs.BlockEpochEvent
...
+		}
+	case <-time.After(15 * time.Second):
+		// TODO: adjust timeout?
+		// The peer was not able to provide us with a response within
+		// our timeout.
+		return nil, fmt.Errorf("did not receive block from peer " +
+			"within timeout of 15 seconds")
+	}
+
+	return nil, nil
```
Comment (halseth) [inline chainntnfs/bitcoindnotify/bitcoind.go]: If we add a default case to cast, I don't think we can ever end up here?

File: chainntnfs/bitcoindnotify/bitcoind.go [hunk 5]
```diff
@@ -716,3 +1161,102 @@ func (b *BitcoindNotifier) RegisterBlockEpochNtfn() (*chainntnfs.BlockEpochEvent
...
+		switch msg.(type) {
+		case *btcutil.Block:
+			return msg.(*btcutil.Block), nil
+		case *wire.MsgNotFound:
+			// The peer was not able to find a block with the
+			// requested hash.
+			return nil, fmt.Errorf("block with hash %v not found",
+				blockHash)
+		}
+	case <-time.After(15 * time.Second):
```
Comment (halseth) [inline chainntnfs/bitcoindnotify/bitcoind.go]: I think we should make this method a method on the `Notifier`, such that we can return in the case `<-notifier.quit`

File: chainntnfs/bitcoindnotify/bitcoind.go [hunk 6]
```diff
@@ -716,3 +1161,102 @@ func (b *BitcoindNotifier) RegisterBlockEpochNtfn() (*chainntnfs.BlockEpochEvent
...
+
+	// Allow the peer to start exchanging messages on the established
+	// connection.
+	p.AssociateConnection(conn)
+
+	// Since the peer will reject all messages without establishing a
+	// version handshake first, we make sure it has succeeded.
+	select {
+	case <-verack:
+	case <-time.After(15 * time.Second):
```
Comment (halseth) [inline chainntnfs/bitcoindnotify/bitcoind.go]: same regarding `<-quit`

File: chainntnfs/btcdnotify/btcd.go [hunk 7]
```diff
@@ -3,20 +3,24 @@ package btcdnotify
 import (
 	"errors"
 	"fmt"
+	"net"
```
Comment (halseth) [inline chainntnfs/btcdnotify/btcd.go]: Since this commit is getting quite big, maybe an idea to split it into two? :)

File: chainntnfs/btcdnotify/btcd.go [hunk 8]
```diff
@@ -32,6 +36,19 @@ var (
...
 	ErrChainNotifierShuttingDown = errors.New("chainntnfs: system interrupt " +
 		"while attempting to register for spend notification.")
+
+	// errTxNotFound is used whenever we try looking for a transaction, but
+	// we weren't able to find it.
+	errTxNotFound = errors.New("transaction not found")
+
+	// netToParams maps the Bitcoin network specific magic number to its
+	// parameters.
+	netToParams = map[wire.BitcoinNet]*chaincfg.Params{
```
Comment (halseth) [inline chainntnfs/btcdnotify/btcd.go]: same

File: chainntnfs/btcdnotify/btcd.go [hunk 9]
```diff
@@ -406,59 +437,496 @@ out:
...
+		p, err := connectToOutboundPeer(info.Addr, netToParams[btcNet],
+			listeners, msgChan)
+		if err != nil {
+			continue
+		}
+
+		// Begin looking for the spending transaction by fetching blocks
+		// from our peer at every height and scanning them.
+		for height := lastHeight; height <= bestHeight; height++ {
+			blockHash, err := b.chainConn.GetBlockHash(int64(height))
```
Comment (halseth) [inline chainntnfs/btcdnotify/btcd.go]: Not sure how easy it will be, but could we extract common code by putting the `b.chainConn` behind an interface that could be used by both `btcd` and `bitcoind``

File: chainntnfs/btcdnotify/btcd.go (hunk 7 above)
Comment (halseth) [reply chainntnfs/btcdnotify/btcd.go]: Not gonna review this file for now, as I reckon it's mostly the same changes as for `bitcoind`. We should really try to dedup some of the code, if possible.

File: chainntnfs/btcdnotify/btcd_test.go [hunk 10]
```diff
@@ -0,0 +1,824 @@
...
+
+	privKey, pubKey = btcec.PrivKeyFromBytes(btcec.S256(), testPrivKey)
+	pubKeyAddr, _   = btcutil.NewAddressPubKey(pubKey.SerializeCompressed(),
+		&chaincfg.RegressionNetParams)
+	testAddr = pubKeyAddr.AddressPubKeyHash()
+)
+
+// testConfDetailsFromTxIndex tests that we can find the transaction from the
+// node's transaction index and that the details retrieved are correct.
+func testConfDetailsFromTxIndex(t *testing.T, miner *rpctest.Harness,
```
Comment (halseth) [inline chainntnfs/btcdnotify/btcd_test.go]: Are these similar to those for `bitcoind`? Possible to put the two notifiers behind a common interface, and reuse the tests?

File: chainntnfs/interface_test.go [hunk 11]
```diff
@@ -15,17 +15,16 @@ import (
 
 	"github.com/lightninglabs/neutrino"
 	"github.com/lightningnetwork/lnd/chainntnfs"
-	"github.com/ltcsuite/ltcd/btcjson"
-	"github.com/roasbeef/btcd/chaincfg/chainhash"
-	"github.com/roasbeef/btcwallet/walletdb"
-
 	"github.com/roasbeef/btcd/btcec"
+	"github.com/roasbeef/btcd/btcjson"
```
Comment (halseth) [inline chainntnfs/interface_test.go]: why was this file changed?

**Comment by wilmerpaulino** [comment]
Thanks for the review @halseth!

This PR has been modified to provide manually scanning the backend's chain as the only fallback method when the transaction index is not enabled.

Before attempting to include support for pruned nodes, the `chainntnfs` package should be heavily refactored to avoid more code duplication.

File: chainntnfs/bitcoindnotify/bitcoind.go [hunk 12]
```diff
@@ -249,23 +249,28 @@ out:
...
 				// active chain.
-				txConf, err := b.historicalConfDetails(msg.TxID)
+				txConf, err := b.historicalConfDetails(
+					msg.TxID, msg.heightHint, currentHeight,
+				)
 				if err != nil {
 					chainntnfs.Log.Error(err)
 				}
-				b.heightMtx.RLock()
+
```
Comment (Roasbeef) [inline chainntnfs/bitcoindnotify/bitcoind.go]: Why is the mutex around access to the `txConfNotifier` no longer needed?

File: chainntnfs/bitcoindnotify/bitcoind.go [hunk 13]
```diff
@@ -385,60 +390,123 @@ out:
...
+// in a block in the active chain by using the backend node's transaction index.
+// If it is found, the confirmation details are returned. Otherwise, an error
+// is not returned.
+func (b *BitcoindNotifier) confDetailsFromTxIndex(txid *chainhash.Hash,
 ) (*chainntnfs.TxConfirmation, error) {
 
-	// If the transaction already has some or all of the confirmations,
-	// then we may be able to dispatch it immediately.
-	// TODO: fall back to scanning blocks if txindex isn't on.
 	tx, err := b.chainConn.GetRawTransactionVerbose(txid)
```
Comment (Roasbeef) [inline chainntnfs/bitcoindnotify/bitcoind.go]: The prior comment, minus the TOOD which will now be fixed, can likely remain.

File: chainntnfs/bitcoindnotify/bitcoind.go [hunk 14]
```diff
@@ -385,60 +390,123 @@ out:
...
 	}
 
 	// As we need to fully populate the returned TxConfirmation struct,
 	// grab the block in which the transaction was confirmed so we can
 	// locate its exact index within the block.
 	blockHash, err := chainhash.NewHashFromStr(tx.BlockHash)
 	if err != nil {
-		return nil, fmt.Errorf("unable to get block hash %v for historical "+
-			"dispatch: %v", tx.BlockHash, err)
+		return nil, err
```
Comment (Roasbeef) [inline chainntnfs/bitcoindnotify/bitcoind.go]: Why was the wrapped error (which had more detail) removed?

File: chainntnfs/bitcoindnotify/bitcoind.go [hunk 15]
```diff
@@ -385,60 +390,123 @@ out:
...
+	// If the block was obtained, locate the transaction's index within the
 	// block so we can give the subscriber full confirmation details.
-	txIndex := -1
 	targetTxidStr := txid.String()
-	for i, txHash := range block.Tx {
+	for txIndex, txHash := range block.Tx {
 		if txHash == targetTxidStr {
-			txIndex = i
-			break
+			return &chainntnfs.TxConfirmation{
```
Comment (Roasbeef) [inline chainntnfs/bitcoindnotify/bitcoind.go]: 👍 on the simplification here

File: chainntnfs/bitcoindnotify/bitcoind.go [hunk 16]
```diff
@@ -385,60 +390,123 @@ out:
...
+	return nil, fmt.Errorf("unable to locate tx %v in block %v", txid,
+		blockHash)
+}
+
+// confDetailsManually looks up whether a transaction is already included in a
+// block in the active chain by scanning the chain's blocks, starting from the
+// earliest height the transaction could have been included in, to the current
+// height in the chain. If it is found, the confirmation details are returned.
+// Otherwise, an error is not returned.
+func (b *BitcoindNotifier) confDetailsManually(txid *chainhash.Hash,
```
Comment (Roasbeef) [inline chainntnfs/bitcoindnotify/bitcoind.go]: With this change, we'll need to do a _clean sweep_ through the _entire_ codebase to ensure that `heightHint` is set  properly. There are some known cases with neutrino (particularly in the `contractcourt`) where an incorrect height hint value can cause us to actually miss a notification all together. We'll need to extensively test this patch to ensure these holes are patched up.

File: chainntnfs/bitcoindnotify/bitcoind.go [hunk 17]
```diff
@@ -508,38 +578,82 @@ func (b *BitcoindNotifier) RegisterSpendNtfn(outpoint *wire.OutPoint,
...
+		if tx != nil && tx.BlockHash != "" {
+			// If we're able to retrieve a valid block hash from the
+			// transaction, then we'll use it as our rescan starting
+			// point.
+			blockHash, err = chainhash.NewHashFromStr(tx.BlockHash)
+			if err != nil {
 				return nil, err
 			}
+		} else {
+			// Otherwise, we'll rely on manually scanning our chain
```
Comment (Roasbeef) [inline chainntnfs/bitcoindnotify/bitcoind.go]: Why's this block necessary? Instead we can just query for the block hash of the height hint. Then the block below can be simplified to simply call into `Rescan` with the proper block hash.

File: chainntnfs/btcdnotify/btcd.go [hunk 18]
```diff
@@ -528,33 +598,78 @@ func (b *BtcdNotifier) RegisterSpendNtfn(outpoint *wire.OutPoint,
...
+			// for the block the transaction was included in,
+			// starting from the earliest height it could have been
+			// included in to the current.
+			_, currentHeight, err := b.chainConn.GetBestBlock()
+			if err != nil {
 				return nil, err
 			}
+
+			targetTxidStr := outpoint.Hash.String()
+		out:
```
Comment (Roasbeef) [inline chainntnfs/btcdnotify/btcd.go]: Similar comment here w.r.t this block not being necessary.

File: chainntnfs/bitcoindnotify/bitcoind.go (hunk 17 above)
Comment (wilmerpaulino) [reply chainntnfs/bitcoindnotify/bitcoind.go]: Good point, will fix.

File: chainntnfs/bitcoindnotify/bitcoind.go (hunk 16 above)
Comment (wilmerpaulino) [reply chainntnfs/bitcoindnotify/bitcoind.go]: I was running this patch on testnet for a bit and didn't run into any issues. I'll continue testing and see if I can run into any more.

File: chainntnfs/bitcoindnotify/bitcoind.go (hunk 12 above)
Comment (wilmerpaulino) [reply chainntnfs/bitcoindnotify/bitcoind.go]: The mutex is only being used to guard the `ChainNotifier`'s `bestHeight`. Calling `txConfNotifier.Register` without it should be fine since it doesn't have an effect on the `bestHeight`. I'm guessing it was needed at some point and it didn't get removed. I could be wrong though.

File: chainntnfs/bitcoindnotify/bitcoind.go (hunk 17 above)
Comment (wilmerpaulino) [reply chainntnfs/bitcoindnotify/bitcoind.go]: Fixed.

File: chainntnfs/btcdnotify/btcd.go (hunk 18 above)
Comment (wilmerpaulino) [reply chainntnfs/btcdnotify/btcd.go]: Fixed.

File: chainntnfs/bitcoindnotify/bitcoind.go (hunk 14 above)
Comment (wilmerpaulino) [reply chainntnfs/bitcoindnotify/bitcoind.go]: Fixed.

File: chainntnfs/bitcoindnotify/bitcoind.go (hunk 13 above)
Comment (wilmerpaulino) [reply chainntnfs/bitcoindnotify/bitcoind.go]: Fixed.

File: chainntnfs/bitcoindnotify/bitcoind.go [hunk 19]
```diff
@@ -385,60 +390,127 @@ out:
 // historicalConfDetails looks up whether a transaction is already included in a
 // block in the active chain and, if so, returns details about the confirmation.
 func (b *BitcoindNotifier) historicalConfDetails(txid *chainhash.Hash,
+	heightHint, currentHeight uint32) (*chainntnfs.TxConfirmation, error) {
+
+	// First, we'll attempt to retrieve the transaction details using the
+	// backend node's transaction index.
+	txConf, err := b.confDetailsFromTxIndex(txid)
+	if err != nil {
```
Comment (wilmerpaulino) [inline chainntnfs/bitcoindnotify/bitcoind.go]: Noticed this while addressing your comments: maybe we should proceed to use the `heightHint` if this fails, rather than returning the error?

Comment (Roasbeef) [reply chainntnfs/bitcoindnotify/bitcoind.go]: With the new code, the nature of the error is _likely_ critical. We'll need to properly test with both `bitcoind` and `btcd` to ensure that our logic catches the "no tx index" error properly.

**Comment by Roasbeef** [comment]
Can you rebase this to master? Thanks!

**Comment by wpaulino** [comment]
Rebased.

File: chainntnfs/btcdnotify/btcd.go [hunk 20]
```diff
@@ -447,59 +450,127 @@ out:
...
+	// Begin scanning blocks at every height to determine where the
+	// transaction was included in.
+	for height := heightHint; height <= currentHeight; height++ {
+		blockHash, err := b.chainConn.GetBlockHash(int64(height))
+		if err != nil {
+			return nil, fmt.Errorf("unable to get hash from block "+
+				"with height %d", height)
+		}
+
+		block, err := b.chainConn.GetBlockVerbose(blockHash)
```
Comment (Roasbeef) [inline chainntnfs/btcdnotify/btcd.go]: Worth adding a `TODO` here that we can actually just fetch the neutrino filters here. Will save us a bit of processing as we can quickly ascertain if a transaction _maybe_ includes a block or not.

**Comment by Roasbeef** [comment]
When we (eventually) port over the integration tests to run on all backends, we should ensure that we're able to start and properly execute the tests against instances of `btcd` and `bitcoind` that don't have the `txindex` active.

File: chainntnfs/bitcoindnotify/bitcoind.go [hunk 21]
```diff
@@ -385,60 +390,127 @@ out:
...
+
+	// If the backend node's transaction index is not enabled, then we'll
+	// fall back to manually scanning the chain's blocks, looking for the
+	// block where the transaction was included in.
+	return b.confDetailsManually(txid, heightHint, currentHeight)
+}
+
+// confDetailsFromTxIndex looks up whether a transaction is already included
+// in a block in the active chain by using the backend node's transaction index.
+// If it is found, the confirmation details are returned. Otherwise, an error
```
Comment (halseth) [inline chainntnfs/bitcoindnotify/bitcoind.go]: Comment describing if/when an error is returned is confusing.

File: chainntnfs/bitcoindnotify/bitcoind.go [hunk 22]
```diff
@@ -385,60 +390,127 @@ out:
...
+		// proceed with fallback methods.
+		jsonErr, ok := err.(*btcjson.RPCError)
+		if !ok || jsonErr.Code != btcjson.ErrRPCNoTxInfo {
+			return nil, fmt.Errorf("unable to query for txid "+
+				"%v: %v", txid, err)
 		}
-		return nil, fmt.Errorf("unable to query for txid(%v): %v", txid, err)
+	}
+
+	// Make sure we actually retrieved a transaction that is included in a
```
Comment (halseth) [inline chainntnfs/bitcoindnotify/bitcoind.go]: *thumbsup* on this refactoring!

File: chainntnfs/bitcoindnotify/bitcoind.go [hunk 23]
```diff
@@ -385,60 +390,127 @@ out:
...
+	// within the block, but didn't.
+	return nil, fmt.Errorf("unable to locate tx %v in block %v", txid,
+		blockHash)
+}
+
+// confDetailsManually looks up whether a transaction is already included in a
+// block in the active chain by scanning the chain's blocks, starting from the
+// earliest height the transaction could have been included in, to the current
+// height in the chain. If it is found, the confirmation details are returned.
+// Otherwise, an error is not returned.
```
Comment (halseth) [inline chainntnfs/bitcoindnotify/bitcoind.go]: Same. I _think_ I understand what this means, but it's a bit confusing 😛

File: chainntnfs/btcdnotify/btcd.go (hunk 20 above)
Comment (wpaulino) [reply chainntnfs/btcdnotify/btcd.go]: Done.

File: chainntnfs/bitcoindnotify/bitcoind.go (hunk 23 above)
Comment (wpaulino) [reply chainntnfs/bitcoindnotify/bitcoind.go]: Fixed.

File: chainntnfs/bitcoindnotify/bitcoind.go (hunk 21 above)
Comment (wpaulino) [reply chainntnfs/bitcoindnotify/bitcoind.go]: Fixed.

**Comment by wpaulino** [comment]
Rebased and added TODOs.

File: chainntnfs/bitcoindnotify/bitcoind.go [hunk 24]
```diff
@@ -367,60 +377,130 @@ func (b *BitcoindNotifier) handleRelevantTx(tx chain.RelevantTx, bestHeight int3
...
+	// If the backend node's transaction index is not enabled, then we'll
+	// fall back to manually scanning the chain's blocks, looking for the
+	// block where the transaction was included in.
+	return b.confDetailsManually(txid, heightHint, currentHeight)
+}
+
+// confDetailsFromTxIndex looks up whether a transaction is already included
+// in a block in the active chain by using the backend node's transaction index.
+// If the transaction is found, its confirmation details are returned.
+// Otherwise, a nil error is returned. A non-nil error may also be returned if
```
Comment (halseth) [inline chainntnfs/bitcoindnotify/bitcoind.go]: Suggestion:
"If the transaction is found in the index, its confirmation details are returned. If it is not found in the index, a nil confirmation detail is returned." 
I think that should make it more clear that a `non-nil` error is _critical_(?).
